From 63a1ea90a0cec749a2c34d4fdb71e16c294c63ed Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Fri, 24 Feb 2023 16:23:54 +0100
Subject: [PATCH 0348/1695] drivers: stm32_exti: support lock of the RIF
 configuration

Lock EXTI resources when configured. This cannot be undone until
next reset.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 434f49165c7eb6cda342b9e4229ff444a78b701c drivers: stm32_exti: support lock of the RIF configuration]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: Iea846026333a1eceb44b8a5ee3ee660257f05020
---
 core/drivers/stm32_exti.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/core/drivers/stm32_exti.c b/core/drivers/stm32_exti.c
index 998b64195..400c1c75d 100644
--- a/core/drivers/stm32_exti.c
+++ b/core/drivers/stm32_exti.c
@@ -24,6 +24,7 @@
 #define _EXTI_SECCFGR(n)	(0x014U + (n) * 0x20U)
 #define _EXTI_PRIVCFGR(n)	(0x018U + (n) * 0x20U)
 #define _EXTI_CR(n)		(0x060U + (n) * 4U)
+#define _EXTI_LOCKR		0x070U
 #define _EXTI_C1IMR(n)		(0x080U + (n) * 0x10U)
 #define _EXTI_EnCIDCFGR(n)	(0x180U + (n) * 4U)
 #define _EXTI_CmCIDCFGR(n)	(0x300U + (n) * 4U)
@@ -35,6 +36,9 @@
 /* PRIVCFGR register bitfields */
 #define _EXTI_PRIVCFGR_MASK		GENMASK_32(31, 0)
 
+/* LOCKR register bitfields */
+#define _EXTI_LOCKR_GLOCK		BIT(0)
+
 /* CIDCFGR register bitfields */
 #define _EXTI_CIDCFGR_SCID_MASK		GENMASK_32(6, 4)
 #define _EXTI_CIDCFGR_CONF_MASK		(_CIDCFGR_CFEN | \
@@ -67,6 +71,8 @@ struct stm32_exti_pdata {
 	uint32_t *c_cids;
 
 	SLIST_ENTRY(stm32_exti_pdata) link;
+
+	bool glock;
 };
 
 static SLIST_HEAD(, stm32_exti_pdata) stm32_exti_list =
@@ -272,6 +278,11 @@ static void stm32_exti_rif_parse_dt(struct stm32_exti_pdata *exti,
 	unsigned int i = 0;
 	int len = 0;
 
+	if (fdt_getprop(fdt, node, "st,glocked", NULL))
+		exti->glock = true;
+	else
+		DMSG("No global lock on RIF configuration");
+
 	cuint = fdt_getprop(fdt, node, "st,protreg", &len);
 	if (!cuint)
 		panic("No RIF configuration available");
@@ -381,6 +392,10 @@ static TEE_Result stm32_exti_rif_apply(const struct stm32_exti_pdata *exti)
 				_EXTI_CIDCFGR_CONF_MASK, exti->c_cids[i]);
 	}
 
+	/* If TDCID, configure global lock */
+	if (exti->glock)
+		io_setbits32(exti->base + _EXTI_LOCKR, _EXTI_LOCKR_GLOCK);
+
 	return TEE_SUCCESS;
 }
 
-- 
2.47.2

